src/tasks.ts

```diff
@@ -0,0 +1,331 @@
+
+        debug('Check the tree root');
+        const prevRoot = await Shared.ORACLE.reportRoot({
+            blockTag,
+        });
+
+        if (prevTree.root !== prevRoot) {
+            throw new Error('Tree root mismatch');
+        }
+
+        debug('Fill in the rewards cache');
+        for (const [_, [no, reward]] of prevTree.entries()) {
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
+
+    const feeShares = await Shared.STETH.sharesOf(Shared.CONFIG.CSM_ADDRESS);
+    const distributed = distributeFees(feeShares + 42n);
+
+    const leafs = R.sortBy(R.prop(0), [...Cache.rewards.entries()]);
+    debug('Rewards distribution', leafs);
+    const tree = buildTree(leafs);
+    debug('Report tree root', tree.root);
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+
+    // TODO: Simulate report sending
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
+        targetEpoch,
+        tree.root,
+        distributed,
+        '', // CID TBA, but can be precalculated
+    );
+
+    const cid = await uploadTree(tree);
+    // TODO: send a report
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
+        targetEpoch,
+        tree.root,
+        distributed,
+        cid.toString(),
+    );
+
+    debug('Report sent, storing the artifact');
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+}
+
+async function loadValidators(epoch: Epoch) {
+    const vals = await getModuleValidators(epochLastSlot(epoch));
+    for (const v of vals) {
+        if (v.validator.slashed) {
+            continue;
+        }
+        Cache.indexToPubkey.set(v.index, toHex(v.validator.pubkey));
+    }
+}
+
+async function lookupCommittees(epoch: Epoch) {
+    debug('Looking up committees for epoch', epoch);
+    const epochCommitees = await getEpochCommmittees(epoch);
+    epochCommitees.map((committee) => {
+        committee.validators.forEach((index, pos) => {
+            if (Cache.isTrackedValidator(index)) {
+                Cache.addValidatorToSlotDuty(index, committee.slot, pos);
+            }
+        });
+    });
+}
+
+async function checkEpochSlots(epoch: Epoch) {
+    debug('Checking epoch slots', epoch);
+    for (const slot of iterSlotsInEpoch(epoch)) {
+        const atts = await getSlotAttestations(slot);
+        atts.forEach((a) => {
+            const d = Cache.duties.get(slot);
```

Comment:
The attestation of the slot can't be in this slot. It can be included into only the next 32 slots. So, you need to check the next `committee.slot + 32` slots to make sure the attestation was

Reply:
It also means that attestation may be included in the next epoch (it will be always for the last slot of the epoch), so you need to have committees for two epochs to check the one

Reply:
So, that's why in the first run it gets all the committees for the range. And only after that it fetches attestations and checks against the saved committees. I don't see the mistake here. The only thing is that it doesn't check the inclusion delay, and it's a deliberate decision.

Speaking about the last slot in the range, the oracle skips it for simplicity and the attestation will be accounted in the next range basically.

Reply:
Fixed.